File: orchestrator/tools/checklist.py
"""Checklist tool for running validation checks in workflows.

Supports three check types:
- bash: Run shell commands and compare output
- regex: Pattern matching in files using ripgrep
- model: LLM-based judgment using Claude haiku

All checks run in parallel for faster execution.
"""

import asyncio
import re
import subprocess
from dataclasses import dataclass
from pathlib import Path
from typing import TYPE_CHECKING, Any, Dict, List, Optional

import yaml

from .base import BaseTool, ToolResult
from ..display_adapter import get_display

if TYPE_CHECKING:
    from ..context import ExecutionContext
    from ..tmux import TmuxManager


@dataclass
class CheckResult:
    """Result of a single check."""

    name: str
    passed: bool
    severity: str  # error, warning, info
    message: str
    details: Optional[str] = None


class ChecklistTool(BaseTool):
    """Execute validation checklists with bash, regex, and model checks."""

    @property
    def name(self) -> str:
        return "checklist"

    def validate_step(self, step: Dict[str, Any]) -> None:
        """Validate checklist step configuration."""
        if "checklist" not in step and "items" not in step:
            raise ValueError(
                "Checklist step requires either 'checklist' (file name) "
                "or 'items' (inline check definitions)"
            )

        # Validate inline items if provided
        if "items" in step:
            items = step["items"]
            if not isinstance(items, list):
                raise ValueError("'items' must be a list of check definitions")
            for i, item in enumerate(items):
                if "name" not in item:
                    raise ValueError(f"Check item {i} missing required 'name' field")
                if "type" not in item:
                    raise ValueError(f"Check item {i} missing required 'type' field")
                if item["type"] not in ("bash", "regex", "model"):
                    raise ValueError(
                        f"Check item {i} has invalid type '{item['type']}'. "
                        "Valid types: bash, regex, model"
                    )

    def execute(
        self,
        step: Dict[str, Any],
        context: "ExecutionContext",
        tmux_manager: "TmuxManager",
    ) -> ToolResult:
        """Execute checklist and return aggregated results."""
        import time

        start_time = time.time()
        display = get_display()

        # Load checklist configuration
        checklist_config = self._load_checklist(step, context)
        if checklist_config is None:
            return ToolResult(
                success=False,
                error="Failed to load checklist configuration",
            )

        checklist_name = checklist_config.get("name", "unnamed")
        items = checklist_config.get("items", [])
        on_fail = step.get("on_fail", checklist_config.get("on_fail", "warn"))

        # Show checklist start
        display.print_checklist_start(checklist_name, len(items))

        # Run all checks in parallel
        # Handle both sync and async contexts
        try:
            loop = asyncio.get_running_loop()
            # Already in async context - create task
            import concurrent.futures
            with concurrent.futures.ThreadPoolExecutor() as pool:
                future = pool.submit(
                    asyncio.run,
                    self._run_checks_parallel(items, context)
                )
                results = future.result()
        except RuntimeError:
            # No running loop - use asyncio.run directly
            results = asyncio.run(self._run_checks_parallel(items, context))

        # Display each check result
        for result in results:
            display.print_checklist_item(
                name=result.name,
                passed=result.passed,
                severity=result.severity,
                message=result.message if not result.passed else None,
                details=result.details,
            )

        # Calculate stats
        passed_count = sum(1 for r in results if r.passed)
        total_count = len(results)
        has_errors = any(
            not r.passed and r.severity == "error" for r in results
        )
        has_warnings = any(
            not r.passed and r.severity == "warning" for r in results
        )

        # Display completion summary
        duration = time.time() - start_time
        display.print_checklist_complete(
            checklist_name=checklist_name,
            passed_count=passed_count,
            total_count=total_count,
            has_errors=has_errors,
            has_warnings=has_warnings,
            duration=duration,
        )

        # Aggregate results for output
        output = self._format_results(checklist_name, results)

        # Success depends on on_fail mode
        if on_fail == "stop":
            success = not has_errors and not has_warnings
        elif on_fail == "warn":
            success = not has_errors  # Warnings don't fail
        else:  # continue
            success = True  # Always succeed

        return ToolResult(
            success=success,
            output=output,
            error=None if success else "Checklist validation failed",
        )

    def _load_checklist(
        self,
        step: Dict[str, Any],
        context: "ExecutionContext",
    ) -> Optional[Dict[str, Any]]:
        """Load checklist from file or inline items."""
        # Inline items take precedence
        if "items" in step:
            return {
                "name": step.get("name", "inline-checklist"),
                "items": step["items"],
                "on_fail": step.get("on_fail", "warn"),
            }

        # Load from file
        checklist_name = context.interpolate(step["checklist"])
        checklist_dir = context.project_path / ".claude" / "checklists"

        # Try with and without .yaml extension
        for ext in ("", ".yaml", ".yml"):
            checklist_path = checklist_dir / f"{checklist_name}{ext}"
            if checklist_path.exists():
                with open(checklist_path, "r") as f:
                    return yaml.safe_load(f)

        # Checklist file not found
        return None

    async def _run_checks_parallel(
        self,
        items: List[Dict[str, Any]],
        context: "ExecutionContext",
    ) -> List[CheckResult]:
        """Run all checks in parallel and return results."""
        tasks = [self._run_check_async(item, context) for item in items]
        return await asyncio.gather(*tasks)

    async def _run_check_async(
        self,
        item: Dict[str, Any],
        context: "ExecutionContext",
    ) -> CheckResult:
        """Run a single check asynchronously and return result."""
        check_type = item["type"]
        check_name = item["name"]
        severity = item.get("severity", "warning")

        try:
            if check_type == "bash":
                return await self._run_bash_check_async(item, context, severity)
            elif check_type == "regex":
                return await self._run_regex_check_async(item, context, severity)
            elif check_type == "model":
                return await self._run_model_check_async(item, context, severity)
            else:
                return CheckResult(
                    name=check_name,
                    passed=False,
                    severity="error",
                    message=f"Unknown check type: {check_type}",
                )
        except Exception as e:
            return CheckResult(
                name=check_name,
                passed=False,
                severity=severity,
                message=f"Check execution failed: {str(e)}",
            )

    async def _run_bash_check_async(
        self,
        item: Dict[str, Any],
        context: "ExecutionContext",
        severity: str,
    ) -> CheckResult:
        """Run a bash command check asynchronously."""
        name = item["name"]
        command = context.interpolate(item["command"])

        try:
            # Run subprocess in thread pool to avoid blocking
            result = await asyncio.to_thread(
                subprocess.run,
                command,
                shell=True,
                cwd=str(context.project_path),
                capture_output=True,
                text=True,
                timeout=60,
            )
            output = result.stdout.strip()

            # Check expectations
            passed = True
            message = ""

            if "expect" in item:
                expected = str(item["expect"])
                if output != expected:
                    passed = False
                    message = f"Expected '{expected}', got '{output}'"
                else:
                    message = "Output matches expected value"

            elif "expect_not" in item:
                forbidden = str(item["expect_not"])
                if forbidden in output:
                    passed = False
                    message = f"Output contains forbidden value: {forbidden}"
                else:
                    message = "Output does not contain forbidden value"

            elif "expect_regex" in item:
                pattern = item["expect_regex"]
                if not re.search(pattern, output):
                    passed = False
                    message = f"Output does not match pattern: {pattern}"
                else:
                    message = "Output matches pattern"

            else:
                # No expectation, just check exit code
                passed = result.returncode == 0
                message = "Command succeeded" if passed else f"Command failed with exit code {result.returncode}"

            return CheckResult(
                name=name,
                passed=passed,
                severity=severity,
                message=message,
                details=output if not passed else None,
            )

        except subprocess.TimeoutExpired:
            return CheckResult(
                name=name,
                passed=False,
                severity=severity,
                message="Command timed out after 60 seconds",
            )

    async def _run_regex_check_async(
        self,
        item: Dict[str, Any],
        context: "ExecutionContext",
        severity: str,
    ) -> CheckResult:
        """Run a regex pattern matching check using ripgrep asynchronously."""
        name = item["name"]
        pattern = item["pattern"]
        files = item.get("files", "**/*.py")
        exclude = item.get("exclude", "")
        expect_count = item.get("expect", 0)

        # Build ripgrep command
        cmd_parts = ["rg", "--count-matches", "-e", pattern]

        # Add file glob
        cmd_parts.extend(["--glob", files])

        # Add exclude patterns
        if exclude:
            for excl in exclude.split(","):
                cmd_parts.extend(["--glob", f"!{excl.strip()}"])

        # Add path
        cmd_parts.append(".")

        try:
            # Run subprocess in thread pool to avoid blocking
            result = await asyncio.to_thread(
                subprocess.run,
                cmd_parts,
                cwd=str(context.project_path),
                capture_output=True,
                text=True,
                timeout=60,
            )

            # Parse ripgrep output to count matches
            # Format: file:count per line
            total_matches = 0
            match_details = []
            for line in result.stdout.strip().split("\n"):
                if line and ":" in line:
                    parts = line.rsplit(":", 1)
                    if len(parts) == 2:
                        try:
                            count = int(parts[1])
                            total_matches += count
                            if count > 0:
                                match_details.append(f"{parts[0]}: {count} matches")
                        except ValueError:
                            pass

            passed = total_matches == expect_count
            if passed:
                message = f"Found {total_matches} matches (expected {expect_count})"
            else:
                message = f"Found {total_matches} matches, expected {expect_count}"

            return CheckResult(
                name=name,
                passed=passed,
                severity=severity,
                message=message,
                details="\n".join(match_details[:10]) if not passed and match_details else None,
            )

        except FileNotFoundError:
            return CheckResult(
                name=name,
                passed=False,
                severity=severity,
                message="ripgrep (rg) not found - please install it",
            )
        except subprocess.TimeoutExpired:
            return CheckResult(
                name=name,
                passed=False,
                severity=severity,
                message="Pattern search timed out after 60 seconds",
            )

    async def _run_model_check_async(
        self,
        item: Dict[str, Any],
        context: "ExecutionContext",
        severity: str,
    ) -> CheckResult:
        """Run an LLM-based check using Claude haiku asynchronously."""
        name = item["name"]
        prompt_template = item["prompt"]
        pass_pattern = item.get("pass_pattern", r"(?i)(PASS|pass|yes|ok|true)")

        # Interpolate variables in prompt
        prompt = context.interpolate(prompt_template)

        # Include any specified context variables
        context_vars = item.get("context_vars", [])
        if context_vars:
            var_context = []
            for var_name in context_vars:
                value = context.get(var_name)
                if value:
                    var_context.append(f"## {var_name}\n{value}")
            if var_context:
                prompt = "\n\n".join(var_context) + "\n\n" + prompt

        try:
            # Already async - no need for asyncio.run
            response = await self._call_haiku(prompt, context)

            # Check if response indicates pass
            passed = bool(re.search(pass_pattern, response))

            return CheckResult(
                name=name,
                passed=passed,
                severity=severity,
                message="Check passed" if passed else "Check failed",
                details=response if not passed else None,
            )

        except Exception as e:
            return CheckResult(
                name=name,
                passed=False,
                severity=severity,
                message=f"Model check failed: {str(e)}",
            )

    async def _call_haiku(
        self,
        prompt: str,
        context: "ExecutionContext",
    ) -> str:
        """Call Claude haiku for model-based checks."""
        try:
            from claude_agent_sdk import ClaudeAgentOptions, query
        except ImportError:
            raise RuntimeError(
                "claude-agent-sdk not installed. Run: pip install claude-agent-sdk"
            )

        options = ClaudeAgentOptions(
            model="claude-3-5-haiku-20241022",
            max_turns=1,
            allowed_tools=[],  # No tools for simple checks
            cwd=str(context.project_path),
        )

        # query() returns an async generator - iterate to get messages
        response_text = ""
        async for message in query(prompt=prompt, options=options):
            # Look for AssistantMessage with text content
            if hasattr(message, "content"):
                for block in message.content:
                    if hasattr(block, "text"):
                        response_text += block.text

        return response_text or "No response"

    def _format_results(
        self,
        checklist_name: str,
        results: List[CheckResult],
    ) -> str:
        """Format check results for output."""
        lines = [f"## Checklist: {checklist_name}"]

        passed_count = sum(1 for r in results if r.passed)
        total_count = len(results)
        warning_count = sum(1 for r in results if not r.passed and r.severity == "warning")
        error_count = sum(1 for r in results if not r.passed and r.severity == "error")

        if error_count > 0:
            status = "FAILED"
        elif warning_count > 0:
            status = "PASSED with warnings"
        else:
            status = "PASSED"

        lines.append(f"Status: {status} ({passed_count}/{total_count} checks passed)")
        if warning_count > 0:
            lines.append(f"Warnings: {warning_count}")
        if error_count > 0:
            lines.append(f"Errors: {error_count}")
        lines.append("")

        # Individual results
        for r in results:
            if r.passed:
                icon = "✓"
            elif r.severity == "error":
                icon = "✗"
            elif r.severity == "warning":
                icon = "⚠"
            else:
                icon = "ℹ"

            lines.append(f"{icon} {r.name}")
            if not r.passed:
                lines.append(f"  {r.message}")
                if r.details:
                    for detail_line in r.details.split("\n")[:5]:
                        lines.append(f"    {detail_line}")

        return "\n".join(lines)

File: ts/src/core/tools/checklist.ts
/**
 * Checklist tool for running validation checks in workflows.
 *
 * Supports three check types:
 * - bash: Run shell commands and compare output
 * - regex: Pattern matching in files using ripgrep
 * - model: LLM-based judgment using Claude haiku
 *
 * All checks run in parallel for faster execution.
 */

import { spawn } from "node:child_process";
import { existsSync, readFileSync } from "node:fs";
import { join } from "node:path";
import Anthropic from "@anthropic-ai/sdk";
import type { StepConfig } from "../../types/index.ts";
import { LoopSignal } from "../../types/index.ts";
import type { ExecutionContext } from "../context/execution.ts";
import type { TmuxManager } from "../tmux/manager.ts";
import type { ToolResult } from "./types.ts";
import { BaseTool, errorResult } from "./types.ts";

/**
 * Result of a single check.
 */
interface CheckResult {
	name: string;
	passed: boolean;
	severity: CheckSeverity;
	message: string;
	details?: string;
}

/**
 * Check item configuration.
 */
interface CheckItem {
	name: string;
	type: "bash" | "regex" | "model";
	severity?: CheckSeverity;
	// bash type
	command?: string;
	expect?: string;
	expectNot?: string;
	expectRegex?: string;
	// regex type
	pattern?: string;
	files?: string;
	exclude?: string;
	// model type
	prompt?: string;
	passPattern?: string;
	contextVars?: string[];
}

/**
 * Checklist configuration (from file or inline).
 */
interface ChecklistConfig {
	name: string;
	items: CheckItem[];
	onFail?: OnFailMode;
}

type CheckSeverity = "error" | "warning" | "info";
type OnFailMode = "stop" | "warn" | "continue";

const MODEL_ALIASES: Record<string, string> = {
	haiku: "claude-haiku-4-5-20250514",
};

/**
 * Execute validation checklists with bash, regex, and model checks.
 */
export class ChecklistTool extends BaseTool {
	private client: Anthropic | null = null;

	get name(): string {
		return "checklist";
	}

	validateStep(step: StepConfig): void {
		const items = step.items as CheckItem[] | undefined;
		const checklistName = step.checklist as string | undefined;

		if (!checklistName && !items) {
			throw new Error(
				"Checklist step requires either 'checklist' (file name) " +
					"or 'items' (inline check definitions)",
			);
		}

		// Validate inline items if provided
		if (items) {
			if (!Array.isArray(items)) {
				throw new Error("'items' must be a list of check definitions");
			}
			for (let i = 0; i < items.length; i++) {
				const item = items[i];
				if (!item.name) {
					throw new Error(`Check item ${i} missing required 'name' field`);
				}
				if (!item.type) {
					throw new Error(`Check item ${i} missing required 'type' field`);
				}
				if (!["bash", "regex", "model"].includes(item.type)) {
					throw new Error(
						`Check item ${i} has invalid type '${item.type}'. ` +
							"Valid types: bash, regex, model",
					);
				}
			}
		}
	}

	async execute(
		step: StepConfig,
		context: ExecutionContext,
		_tmuxManager: TmuxManager,
	): Promise<ToolResult> {
		const startTime = Date.now();

		// Load checklist configuration
		const checklistConfig = this.loadChecklist(step, context);
		if (!checklistConfig) {
			return errorResult("Failed to load checklist configuration");
		}

		const checklistName = checklistConfig.name;
		const items = checklistConfig.items;
		const onFail: OnFailMode =
			(step.onFail as OnFailMode) ?? checklistConfig.onFail ?? "warn";

		// Run all checks in parallel
		const results = await Promise.all(
			items.map((item) => this.runCheck(item, context)),
		);

		// Calculate stats for success determination
		const hasErrors = results.some((r) => !r.passed && r.severity === "error");
		const hasWarnings = results.some(
			(r) => !r.passed && r.severity === "warning",
		);

		// Format output
		const duration = (Date.now() - startTime) / 1000;
		const output = this.formatResults(checklistName, results, duration);

		// Success depends on onFail mode
		let success: boolean;
		if (onFail === "stop") {
			success = !hasErrors && !hasWarnings;
		} else if (onFail === "warn") {
			success = !hasErrors; // Warnings don't fail
		} else {
			success = true; // continue: always succeed
		}

		return {
			success,
			output,
			error: success ? undefined : "Checklist validation failed",
			loopSignal: LoopSignal.NONE,
		};
	}

	private loadChecklist(
		step: StepConfig,
		context: ExecutionContext,
	): ChecklistConfig | null {
		const items = step.items as CheckItem[] | undefined;
		const checklistName = step.checklist as string | undefined;

		// Inline items take precedence
		if (items) {
			return {
				name: step.name ?? "inline-checklist",
				items: items,
				onFail: (step.onFail as OnFailMode) ?? "warn",
			};
		}

		// Load from file
		if (!checklistName) {
			return null;
		}

		const interpolatedName = context.interpolate(checklistName);
		const checklistDir = join(context.projectPath, ".cw", "checklists");

		// Try different extensions
		const extensions = ["", ".json"];
		for (const ext of extensions) {
			const checklistPath = join(checklistDir, `${interpolatedName}${ext}`);
			if (existsSync(checklistPath)) {
				try {
					const content = readFileSync(checklistPath, "utf-8");
					const config = JSON.parse(content) as ChecklistConfig;
					return config;
				} catch (e) {
					const message = e instanceof Error ? e.message : String(e);
					throw new Error(`Failed to parse checklist file: ${message}`);
				}
			}
		}

		return null;
	}

	private async runCheck(
		item: CheckItem,
		context: ExecutionContext,
	): Promise<CheckResult> {
		const checkType = item.type;
		const checkName = item.name;
		const severity = item.severity ?? "warning";

		try {
			switch (checkType) {
				case "bash":
					return await this.runBashCheck(item, context, severity);
				case "regex":
					return await this.runRegexCheck(item, context, severity);
				case "model":
					return await this.runModelCheck(item, context, severity);
				default:
					return {
						name: checkName,
						passed: false,
						severity: "error",
						message: `Unknown check type: ${checkType}`,
					};
			}
		} catch (e) {
			const message = e instanceof Error ? e.message : String(e);
			return {
				name: checkName,
				passed: false,
				severity,
				message: `Check execution failed: ${message}`,
			};
		}
	}

	private async runBashCheck(
		item: CheckItem,
		context: ExecutionContext,
		severity: CheckSeverity,
	): Promise<CheckResult> {
		const name = item.name;
		const command = context.interpolate(item.command ?? "");

		return new Promise((resolve) => {
			const proc = spawn("bash", ["-c", command], {
				cwd: context.projectPath,
				timeout: 60000,
				env: process.env,
			});

			let stdout = "";
			let stderr = "";

			proc.stdout?.on("data", (data: Buffer) => {
				stdout += data.toString();
			});

			proc.stderr?.on("data", (data: Buffer) => {
				stderr += data.toString();
			});

			proc.on("close", (code) => {
				const output = stdout.trim();

				// Check expectations
				let passed = true;
				let message = "";

				if (item.expect !== undefined) {
					const expected = String(item.expect);
					if (output !== expected) {
						passed = false;
						message = `Expected '${expected}', got '${output}'`;
					} else {
						message = "Output matches expected value";
					}
				} else if (item.expectNot !== undefined) {
					const forbidden = String(item.expectNot);
					if (output.includes(forbidden)) {
						passed = false;
						message = `Output contains forbidden value: ${forbidden}`;
					} else {
						message = "Output does not contain forbidden value";
					}
				} else if (item.expectRegex !== undefined) {
					const pattern = item.expectRegex;
					try {
						const regex = new RegExp(pattern);
						if (!regex.test(output)) {
							passed = false;
							message = `Output does not match pattern: ${pattern}`;
						} else {
							message = "Output matches pattern";
						}
					} catch {
						passed = false;
						message = `Invalid regex pattern: ${pattern}`;
					}
				} else {
					// No expectation, just check exit code
					passed = code === 0;
					message = passed
						? "Command succeeded"
						: `Command failed with exit code ${code}`;
				}

				resolve({
					name,
					passed,
					severity,
					message,
					details: !passed ? output || stderr : undefined,
				});
			});

			proc.on("error", (err) => {
				resolve({
					name,
					passed: false,
					severity,
					message: `Command execution failed: ${err.message}`,
				});
			});
		});
	}

	private async runRegexCheck(
		item: CheckItem,
		context: ExecutionContext,
		severity: CheckSeverity,
	): Promise<CheckResult> {
		const name = item.name;
		const pattern = item.pattern ?? "";
		const files = item.files ?? "**/*.ts";
		const exclude = item.exclude ?? "";
		const expectCount = typeof item.expect === "number" ? item.expect : 0;

		// Build ripgrep command
		const cmdParts = ["rg", "--count-matches", "-e", pattern];

		// Add file glob
		cmdParts.push("--glob", files);

		// Add exclude patterns
		if (exclude) {
			for (const excl of exclude.split(",")) {
				cmdParts.push("--glob", `!${excl.trim()}`);
			}
		}

		// Add path
		cmdParts.push(".");

		return new Promise((resolve) => {
			const proc = spawn(cmdParts[0], cmdParts.slice(1), {
				cwd: context.projectPath,
				timeout: 60000,
				env: process.env,
			});

			let stdout = "";

			proc.stdout?.on("data", (data: Buffer) => {
				stdout += data.toString();
			});

			proc.on("close", () => {
				// Parse ripgrep output to count matches
				// Format: file:count per line
				let totalMatches = 0;
				const matchDetails: string[] = [];
				for (const line of stdout.trim().split("\n")) {
					if (line?.includes(":")) {
						const parts = line.split(":");
						const countStr = parts[parts.length - 1];
						const count = Number.parseInt(countStr, 10);
						if (!Number.isNaN(count)) {
							totalMatches += count;
							if (count > 0) {
								matchDetails.push(
									`${parts.slice(0, -1).join(":")}: ${count} matches`,
								);
							}
						}
					}
				}

				const passed = totalMatches === expectCount;
				const message = passed
					? `Found ${totalMatches} matches (expected ${expectCount})`
					: `Found ${totalMatches} matches, expected ${expectCount}`;

				resolve({
					name,
					passed,
					severity,
					message,
					details:
						!passed && matchDetails.length > 0
							? matchDetails.slice(0, 10).join("\n")
							: undefined,
				});
			});

			proc.on("error", (err) => {
				if (err.message.includes("ENOENT")) {
					resolve({
						name,
						passed: false,
						severity,
						message: "ripgrep (rg) not found - please install it",
					});
				} else {
					resolve({
						name,
						passed: false,
						severity,
						message: `Pattern search failed: ${err.message}`,
					});
				}
			});
		});
	}

	private async runModelCheck(
		item: CheckItem,
		context: ExecutionContext,
		severity: CheckSeverity,
	): Promise<CheckResult> {
		const name = item.name;
		const promptTemplate = item.prompt ?? "";
		const passPattern = item.passPattern ?? "(?i)(PASS|pass|yes|ok|true)";

		// Interpolate variables in prompt
		let prompt = context.interpolate(promptTemplate);

		// Include any specified context variables
		const contextVars = item.contextVars ?? [];
		if (contextVars.length > 0) {
			const varContext: string[] = [];
			for (const varName of contextVars) {
				const value = context.get(varName);
				if (value) {
					varContext.push(`## ${varName}\n${String(value)}`);
				}
			}
			if (varContext.length > 0) {
				prompt = `${varContext.join("\n\n")}\n\n${prompt}`;
			}
		}

		try {
			const response = await this.callHaiku(prompt);

			// Check if response indicates pass
			const regex = new RegExp(passPattern, "i");
			const passed = regex.test(response);

			return {
				name,
				passed,
				severity,
				message: passed ? "Check passed" : "Check failed",
				details: !passed ? response : undefined,
			};
		} catch (e) {
			const message = e instanceof Error ? e.message : String(e);
			return {
				name,
				passed: false,
				severity,
				message: `Model check failed: ${message}`,
			};
		}
	}

	private async callHaiku(prompt: string): Promise<string> {
		// Lazy initialize Anthropic client
		if (!this.client) {
			this.client = new Anthropic();
		}

		const model = MODEL_ALIASES.haiku;

		const response = await this.client.messages.create({
			model,
			max_tokens: 1024,
			messages: [{ role: "user", content: prompt }],
		});

		// Extract text content
		const textContent = response.content.find((c) => c.type === "text");
		return textContent?.type === "text" ? textContent.text : "No response";
	}

	private formatResults(
		checklistName: string,
		results: CheckResult[],
		duration: number,
	): string {
		const lines: string[] = [`## Checklist: ${checklistName}`];

		const passedCount = results.filter((r) => r.passed).length;
		const totalCount = results.length;
		const warningCount = results.filter(
			(r) => !r.passed && r.severity === "warning",
		).length;
		const errorCount = results.filter(
			(r) => !r.passed && r.severity === "error",
		).length;

		let status: string;
		if (errorCount > 0) {
			status = "FAILED";
		} else if (warningCount > 0) {
			status = "PASSED with warnings";
		} else {
			status = "PASSED";
		}

		lines.push(
			`Status: ${status} (${passedCount}/${totalCount} checks passed)`,
		);
		if (warningCount > 0) {
			lines.push(`Warnings: ${warningCount}`);
		}
		if (errorCount > 0) {
			lines.push(`Errors: ${errorCount}`);
		}
		lines.push(`Duration: ${duration.toFixed(2)}s`);
		lines.push("");

		// Individual results
		for (const r of results) {
			let icon: string;
			if (r.passed) {
				icon = "\u2713"; // checkmark
			} else if (r.severity === "error") {
				icon = "\u2717"; // X
			} else if (r.severity === "warning") {
				icon = "\u26a0"; // warning sign
			} else {
				icon = "\u2139"; // info
			}

			lines.push(`${icon} ${r.name}`);
			if (!r.passed) {
				lines.push(`  ${r.message}`);
				if (r.details) {
					for (const detailLine of r.details.split("\n").slice(0, 5)) {
						lines.push(`    ${detailLine}`);
					}
				}
			}
		}

		return lines.join("\n");
	}
}
